mangaki/mangaki/admin.py

```diff
@@ -229,60 +230,99 @@ def make_nsfw(self, request, queryset):
 
     make_nsfw.short_description = "Rendre NSFW les œuvres sélectionnées"
 
-    # FIXME : https://github.com/mangaki/mangaki/issues/205
     def update_tags_via_anidb(self, request, queryset):
-        if request.POST.get("post"):
-            kept_ids = request.POST.getlist('checks')
-            for anime_id in kept_ids:
-                anime = Work.objects.get(id=anime_id)
+        works = queryset.all()
 
-                retrieved_tags = anime.retrieve_tags(client)
-                deleted_tags = retrieved_tags["deleted_tags"]
-                added_tags = retrieved_tags["added_tags"]
-                updated_tags = retrieved_tags["updated_tags"]
+        if request.POST.get('confirm'): # Updating tags has been confirmed
+            to_update_work_ids = set(map(int, request.POST.getlist('to_update_work_ids')))
+            nb_updates = len(to_update_work_ids)
 
-                tags = deleted_tags
-                tags.update(added_tags)
-                tags.update(updated_tags)
+            work_ids = list(map(int, request.POST.getlist('work_ids')))
 
-                client.update_tags(anime, tags)
+            tag_titles = request.POST.getlist('tag_titles')
+            tag_weights = list(map(int, request.POST.getlist('weights')))
+            tag_anidb_tag_ids = list(map(int, request.POST.getlist('anidb_tag_ids')))
+            tags = list(map(AniDBTag, tag_titles, tag_weights, tag_anidb_tag_ids))
 
-            self.message_user(request, "Modifications sur les tags faites")
-            return None
+            # Checkboxes to know which tags have to be kept regardless of their pending status
+            tag_checkboxes = request.POST.getlist('tag_checkboxes')
+            tags_to_process = [tuple(map(int, tag_checkbox.split(':'))) for tag_checkbox in tag_checkboxes]
```

Comment:
What about this instead?
```
tags_to_process = [tuple(map(int, tag_checkbox.split(':')))) for tag_checkbox in tag_checkboxes]
tags_to_process = {'{},{}'.format(work_id, tag_aid): (work_id, tag_aid) for work_id, tag_aid in tags_to_process}
```
Resulting in `O(N)` construction of the dict where `N = len(tag_checkboxes)`.

Then, checking if a couple `(work_id, tag_aid)` is to be processed can be done:
`('{},{}'.format(work_id, tag_aid)) in tags_to_process` and results in a `O(1)` lookup.

Resulting in a total algorithm (starting from line 253) of `O(len(work_ids))` rather than `O(len(work_ids) * len(tags_to_process))` (due to line 258 which is a `O(len(tags_to_process))` lookup).

Of course, using hash maps / dicts is not always a solution, sometimes, you'll have to do tradeoffs on memory consumption versus lookup time and it depends also highly on your algorithm, here, it looks like an easy optimization. I think we could do more by using more advanced data structures, but it'd be unnecessary, as long as we have no reason to further optimize this part.